test/opt/test_utils.cpp

```diff
@@ -55,4 +55,48 @@ TEST(JoinNonDebugInsts, Cases) {
                  "the only remaining string"}));
 }
 
+namespace {
+typedef struct {
+  const char* orig_str_;
+  const char* find_substr_;
+  const char* replace_substr_;
+  const char* expected_str_;
+  bool replace_should_succeed_;
+} SubstringReplacementTestCase;
+}
+using ReplaceSubstringInPlaceTest =
+    ::testing::TestWithParam<SubstringReplacementTestCase>;
+
+TEST_P(ReplaceSubstringInPlaceTest, SubstringReplacement) {
+  auto process = std::string(GetParam().orig_str_);
+  ASSERT_STREQ(GetParam().orig_str_, process.c_str());
+  EXPECT_EQ(GetParam().replace_should_succeed_,
+            ReplaceSubstringInPlace(&process, GetParam().find_substr_,
+                                    GetParam().replace_substr_))
+      << "Original string: " << GetParam().orig_str_
+      << " replace: " << GetParam().find_substr_
+      << " to: " << GetParam().replace_substr_
+      << " should returns: " << GetParam().replace_should_succeed_;
+  EXPECT_STREQ(GetParam().expected_str_, process.c_str())
+      << "Original string: " << GetParam().orig_str_
+      << " replace: " << GetParam().find_substr_
+      << " to: " << GetParam().replace_substr_
+      << " expected string: " << GetParam().expected_str_;
+}
+
+INSTANTIATE_TEST_CASE_P(
+    SubstringReplacement, ReplaceSubstringInPlaceTest,
+    ::testing::ValuesIn(std::vector<SubstringReplacementTestCase>(
+        {// orig string, find substring, replace substring, expected string,
+         // replacement happened
+         {"", "", "", "", false},
+         {"", "", "b", "", false},
+         {"", "a", "b", "", false},
+         {"a", "a", "b", "b", true},
+         {"ab", "a", "b", "bb", true},
+         {"abc", "ab", "bc", "bcc", true},
+         {"abc", "ab", "", "c", true},
+         {"abc", "a", "123", "123bc", true},
+         {"abc", "ab", "a", "ac", true},
+         {"abc", "a", "aab", "aabbc", true}})));
```

Comment:
All your cases with non-empty original string actually find the string.  At least one should not.
Also, they always find it in the first position.  At least one should find it elsewhere.